bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.config.xml;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.osgi.framework.Bundle;
+
+/**
+ * Common base class for XML based providers.
+ *
+ * @author Simon Kaufmann - initial contribution, factored out of subclasses
+ *
+ * @param <T_ID> the key type, e.g. ThingTypeUID, ChannelUID, URI,...
+ * @param <T_OBJECT> the object type, e.g. ThingType, ChannelType, ConfigDescription,...
+ */
+public abstract class AbstractXmlBasedProvider<T_ID, T_OBJECT> {
+
+    private class LocalizedKey {
+        public T_ID uid;
+        public String locale;
+
+        public LocalizedKey(T_ID uid, String locale) {
+            this.uid = uid;
+            this.locale = locale;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + getOuterType().hashCode();
+            result = prime * result + ((locale == null) ? 0 : locale.hashCode());
+            result = prime * result + ((uid == null) ? 0 : uid.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            LocalizedKey other = (LocalizedKey) obj;
+            if (!getOuterType().equals(other.getOuterType())) {
```

Comment:
The class is private and as fair as I could see an instance is only created internally. Do you really need to check if the key itself is owned by different objects?